finetuning-llama-7b-on-l4/README.md

```diff
@@ -0,0 +1,381 @@
+
+tokenizer = AutoTokenizer.from_pretrained(model_path, local_files_only=True)
+model = AutoModelForCausalLM.from_pretrained(
+            model_path, torch_dtype=torch.float16, device_map="auto", trust_remote_code=True)
+
+dataset = load_dataset("dell-research-harvard/AmericanStories",
+    "subset_years",
+    year_list=["1809", "1810", "1811", "1812", "1813", "1814", "1815"]
+)
+dataset = concatenate_datasets(dataset.values())
+
+if tokenizer.pad_token is None:
+    tokenizer.add_special_tokens({'pad_token': '[PAD]'})
+    model.resize_token_embeddings(len(tokenizer))
+
+data = dataset.map(lambda x: tokenizer(
+    x["article"], padding='max_length', truncation=True))
+
+lora_config2 = LoraConfig(
+ r=16,
+ lora_alpha=32,
+ lora_dropout=0.05,
+ bias="none",
+ task_type="CAUSAL_LM"
+)
+
+model = prepare_model_for_kbit_training(model)
+
+# add LoRA adaptor
+model = get_peft_model(model, lora_config2)
+model.print_trainable_parameters()
+
+training_args = TrainingArguments(
+        per_device_train_batch_size=1,
+        gradient_accumulation_steps=4,
+        warmup_steps=2,
+        num_train_epochs=1,
+        learning_rate=2e-4,
+        fp16=True,
+        logging_steps=1,
+        output_dir=finetuned_model_path,
+        optim="paged_adamw_32bit",
+)
+
+trainer = transformers.Trainer(
+    model=model,
+    train_dataset=data,
+    args=training_args,
+    data_collator=transformers.DataCollatorForLanguageModeling(tokenizer, mlm=False),
+)
+model.config.use_cache = False  # silence the warnings. Please re-enable for inference!
+
+trainer.train()
+
+# Merge the fine tuned layer with the base model and save it
+model = model.merge_and_unload()
+
+model.save_pretrained(finetuned_model_path)
+tokenizer.save_pretrained(finetuned_model_path)
+# Beginning of story in the dataset
+prompt = """
+In the late action between Generals
+
+
+Brown and Riall, it appears our men fought
+with a courage and perseverance, that would
+"""
+input_ids = tokenizer(prompt, return_tensors="pt").input_ids
+gen_tokens = model.generate(
+    input_ids,
+    do_sample=True,
+    temperature=0.8,
+    max_length=100,
+)
+print(tokenizer.batch_decode(gen_tokens)[0])
+```
+
+Let’s review the high level of what we’ve included in the Fine Tuning file above. First we load the base model from GCS using GCS Fuse. Then we load the dataset from HuggingFace. The finetuning uses PEFT which stands for Parameter-Efficient Fine-Tuning. It is a technique that allows you to fine tune an LLM using a smaller number of parameters, which makes it more efficient, flexible and less computationally expensive.
+
+Next, create a container image that will run <code>[finetune.py](finetune.py)</code>. Create a file named <code>Dockerfile</code> with the following content:
+```Dockerfile
+FROM nvidia/cuda:12.2.0-runtime-ubuntu22.04
+
+RUN apt-get update && \
+    apt-get -y --no-install-recommends install python3 python3-pip git && \
+    rm -rf /var/lib/apt/lists/*
```

Comment:
Is there a hugging face docker image that would work instead? Maybe https://hub.docker.com/r/huggingface/peft-gpu?

Reply:
that would make this guide a lot cleaner, we could just directly run the code from a configmap and no need to build a new image: https://github.com/huggingface/peft/blob/main/docker/peft-gpu/Dockerfile

lets chat tomorrow about this